README.md
# segadev

This repository contains a set of tools intended to ease development
of software for the Sega Genesis (Mega Drive) console.  Currently,
the following programs are included:

* mbitpad
* mdchksum
* mdconvert


## mbitpad

The `mbitpad` program simply appends a given byte (0xFF by default)
to its input until its size is at least a given number of megabits.
For a specific example,

    mbitpad -c 0xFF -s 4 rom.bin

pads the file `rom.bin` with an <FF> character
until it is 4Mbit in size (equivalent to 0.5 MB).
If this file is already at least this large, it is unchanged.
For more information, see

    mbitpad -h


## mdchksum

Correct the checksum in a non-interleaved ROM.
By default, the program merely computes and displays
the checksum that the given file should have,
which is a non-destructive operation.
If the `-f` switch is given, or `-w value`,
then by default a copy of the input file is written to the standard output
except that the location that corresponds to a ROM file's checksum field
is overwritten by either a computed checksum in the former case,
or the given value in the latter.
This is also a non-destructive operation,
though the `-i` switch requests an in-place edit.

For more information, see

    mdchksum -h

or

    man mdchksum


## mdconvert

This program consists of a driver script `mdconvert`
for a few different sub-programs.
See mdconvert/README for more information.


mdchksum/NEWS
1.2:
  * Functionally identical to 1.0.
  * Wider operating system support --- currently
    - Linux
    - Solaris (NEW)

1.0:
  * Reverted from autotools.
  * Added manual page.
  * Changed options.
  * Standard input and output now supported.

0.2:
  * Removed unnecessary dependencies.
  * Changed function linkage to internal.
  * Minor style changes.

0.1:
  * First release.


mdchksum/mdchksum.c
/*****************************************************************************
 * mdchksum: read and fix checksums on Sega Genesis / Mega Drive roms
 * See the file ``COPYING'' for license information.
 *************************************************************************** */
#define _POSIX_C_SOURCE 200112L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

enum {
	E_SUCCESS   = 0,
	E_USAGE     = 1,
	E_NXFILE    = 2,
	E_OUTPUT    = 3,
	E_READ      = 4,
	E_NOMEM     = 5
};
enum {
	M_CALC   = 1,
	M_READ   = 2,
	M_WRITE  = 4,
	M_FIX    = 8
};

#define CHECKSUM_LOCATION    0x18e
#define CHECKSUM_SIZE        0x002
#define DATA_START           0x200
#define SIZE_FIELD_LOCATION  0x1a4
#define SIZE_FIELD_SIZE      0x004
#define ROM_HEADER_SIZE      0x200

static unsigned int calculate_checksum(unsigned char const * const restrict, size_t const);
static unsigned int find_stored_checksum(unsigned char const * const restrict, size_t const);
static void fix_checksum(unsigned char * const restrict, size_t const, unsigned int const);
static void print_help(void);
static void print_version(void);

int
main(int argc, char* argv[])
{
	/* buffer to store rom header and complete ROM */
	unsigned char rom_header[ROM_HEADER_SIZE];
	unsigned char *rom;
	/* input file name, used as output for in-place operations */
	char *fname      = NULL;
	/* option character */
	int c;
	/* an error has occurred */
	int errflag      = 0;
	/* help is requested */
	int hflag        = 0;
	/* loop counter */
	int i;
	/* in-place operation */
	int iflag        = 0;
	/* operation to perform */
	int mode         = 0;
	/* length of ROM file in bytes, as read from the header */
	int rom_size     = 0;
	/* want version info */
	int vflag        = 0;
	/* data to write in explicit mode */
	int wnum         = 0;

	/* We do not use stderr
	 * */
	fclose(stderr);

	while ((c = getopt(argc, argv, ":cfirVw:")) != -1)
	{
		switch (c)
		{
		case 'c':
			if ((mode |= M_CALC) & ~M_CALC)
			{
				errflag = 1;
			}
			break;
		case 'f':
			if ((mode |= M_FIX) & ~M_FIX)
			{
				errflag = 1;
			}
			break;
		case 'i':
			iflag = 1;
			break;
		case 'r':
			if ((mode |= M_READ) & ~M_READ)
			{
				errflag = 1;
			}
			break;
		case 'V':
			vflag = 1;
			break;
		case 'w':
			if ((mode |= M_WRITE) & ~M_WRITE)
			{
				errflag = 1;
			}
			sscanf(optarg, "%i", &wnum);
			break;
		case '?':
		default:
			hflag = 1;
			errflag = (optopt != '?');
			break;
		}
	}

	if (hflag | errflag)
	{
		print_help();
		exit(errflag ? E_USAGE : E_SUCCESS);
	}
	if (vflag)
	{
		print_version();
		exit(E_SUCCESS);
	}
	if (!mode)
	{
		mode = M_CALC;  /* Default mode is "calculate" */
	}
	if ((optind < argc) && (strncmp(argv[optind], "-", 2) != 0))
	{
		fname = argv[optind];
	}
	else
	{
		fname = NULL;
	}
	if (freopen(fname, "rb", stdin) == NULL)
	{
		exit(E_NXFILE);
	}

	/* Read the ROM header into memory */
	if (fread(rom_header, 1, ROM_HEADER_SIZE, stdin) < ROM_HEADER_SIZE)
	{
		exit(E_READ);
	}
	for (i = 0; i < SIZE_FIELD_SIZE; ++i)
	{
		rom_size *= 256;
		rom_size += rom_header[SIZE_FIELD_LOCATION + i];
	}
	rom = calloc(rom_size, 1);
	if (rom == NULL)
	{
		exit(E_NOMEM);
	}
	memcpy(rom, rom_header, ROM_HEADER_SIZE);
	if (fread(rom+ROM_HEADER_SIZE, 1, rom_size - ROM_HEADER_SIZE, stdin)
	    < rom_size - ROM_HEADER_SIZE)
	{
		exit(E_READ);
	}
	/* If we are here, then `rom' contains the ROM image
	 * and we assume it is safe to proceed.
	 * We are also done reading, so we can close stdin.
	 * */
	fclose(stdin);

	/* If not inplace OR not editing the file,
	 * discard output information
	 * */
	if (!iflag || !(mode & (M_FIX | M_WRITE)))
	{
		fname = NULL;
	}
	if (freopen(fname, "wb", stdout) == NULL)
	{
		exit(E_OUTPUT);
	}

	/* Perform operations on ROM data
	 * */
	switch (mode)
	{
	case M_CALC:
		printf("0x%04x\n", calculate_checksum(rom, rom_size));
		break;
	case M_FIX:
		fix_checksum(rom, rom_size, calculate_checksum(rom, rom_size));
		break;
	case M_READ:
		printf("0x%04x\n", find_stored_checksum(rom, rom_size));
		break;
	case M_WRITE:
		fix_checksum(rom, rom_size, wnum);
		break;
	default:
		break;
	}

	if (mode & (M_WRITE | M_FIX))
	{
		fwrite(rom, 1, rom_size, stdout);
	}
	fclose(stdout);

	exit(E_SUCCESS);
}

static unsigned int
calculate_checksum(unsigned char const * const restrict buffer,
                   size_t const size)
{
	unsigned int checksum = 0;
	int i;

	for (i = DATA_START; i < size; i += 2)
	{
		checksum += buffer[i] * 256;
		checksum += buffer[i + 1];
		checksum %= 65536;
	}

	return checksum;
}

static unsigned int
find_stored_checksum(unsigned char const * const restrict buffer,
                     size_t const size)
{
	unsigned int stored_sum = 0;
	int i;

	if (size < CHECKSUM_LOCATION + CHECKSUM_SIZE)
	{
		/* checksum location is not in buffer */
		return 0;
	}

	/* Read big-endian number */
	for (i = CHECKSUM_LOCATION;
	     i < CHECKSUM_LOCATION + CHECKSUM_SIZE;
	     ++i)
	{
		stored_sum *= 256;
		stored_sum += buffer[i];
	}
	return stored_sum;
}

static void
fix_checksum(unsigned char * const restrict buffer,
             size_t size,
             unsigned int const checksum)
{
	int i;
	unsigned int cksum = checksum;

	if (size < CHECKSUM_LOCATION + CHECKSUM_SIZE)
	{
		/* checksum location is not in buffer */
		return;
	}
	/* Write big-endian number */
	for (i = CHECKSUM_SIZE - 1; i >= 0; --i)
	{
		buffer[CHECKSUM_LOCATION + i] = cksum % 256;
		cksum /= 256;
	}
}

static void
print_help(void)
{
	puts("Usage: mdchksum [-c|-f|-r|-w num] [-i] [-?V] [file]");
	puts("Read and write SEGA Genesis / Mega Drive ROM checksums.");
	puts("");
	puts("  -c     : compute checksum.");
	puts("  -f     : fix checksum.");
	puts("  -i     : operate in-place.");
	puts("  -r     : read stored checksum.");
	puts("  -w num : overwrite the checksum with num.");
	puts("  -?     : display this message.");
	puts("  -V     : display version information.");
}

static void
print_version(void)
{
	puts("mdchksum 1.2");
}


mdconvert/s128k/s128k.c
/*****************************************************************************
 * s128k: split a rom to Sega's ideal format:
 * The rom is split into 128 kB odd and even pieces
 * +---------+---------+
 * | 0: Even | 1: Odd  |
 * +---------+---------+
 * | 2: Even | 3: Odd  |
 * +---------+---------+
 * |        ...        |
 * +---------+---------+
 * See the function ``print_license'' below for license information.
 ****************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <errno.h>

#define kB * 1024
#define PERR(str) fprintf(stderr, str)

void s128k(char*);
void print_help();
void print_license();

int
main(int argc, char* argv[])
{
	int ch = 0, hflag = 0, lflag = 0;
	while ( (ch = getopt(argc, argv, "h?l")) != -1 )
	{
		switch (ch)
		{
		case 'h':
		case '?':
			hflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		default:
			hflag = 1;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	
	if (lflag)
	{
		print_license();
	}

	if (hflag || (argc != 1 && !lflag))
	{
		print_help();
	}
	else
	{
		if (argc == 1)
		{
			s128k(argv[0]);
		}
	}
	return EXIT_SUCCESS;
}


void
s128k(char* filename)
{
	ssize_t cpos = 0, length = strlen(filename) + 4;
	char next_even, next_odd;
	char name_buffer[length];
	FILE *ifile = NULL, *odd_file = NULL, *even_file = NULL;
	
	ifile = fopen(filename, "rb");
	if (ifile == NULL)
	{
		err(errno, "Could not open file %s", filename);
		exit(EXIT_FAILURE);
	}
	
	next_even = fgetc(ifile);
	next_odd = fgetc(ifile);
	while (!feof(ifile) && !ferror(ifile)
	       && (odd_file == NULL || !ferror(odd_file))
	       && (even_file == NULL || !ferror(even_file)))
	{
		if (cpos % (128 kB) == 0)
		{
			if (odd_file != NULL) fclose(odd_file);
			if (even_file != NULL) fclose(even_file);
			snprintf(name_buffer, length, "%s.%d",
			         filename, 2 * cpos / (128 kB));
			even_file = fopen(name_buffer, "wb");
			if (even_file == NULL)
			{
				fclose(ifile);
				exit(EXIT_FAILURE);
			}
			
			snprintf(name_buffer, length, "%s.%d",
			         filename, 2 * cpos / (128 kB) + 1);
			odd_file = fopen(name_buffer, "wb");
			if (odd_file == NULL)
			{
				fclose(ifile);
				fclose(even_file);
				exit(EXIT_FAILURE);
			}
		}
		
		fputc(next_odd, odd_file);
		next_even = fgetc(ifile);
		fputc(next_even, even_file);
		next_odd = fgetc(ifile);
		++cpos;
	}
}


void
print_help()
{
	PERR("s128k Copyright (c) 2014, Dakotah Lambert.\n");
	PERR("All rights reserved.\n\n");
	PERR("usage: s128k [-h?l] infile\n\n");
	PERR("  -h, -? : show this message\n");
	PERR("  -l     : display license information\n");
	PERR("  infile : name of file to split\n\n");
}

void
print_license()
{
	PERR("\n s128k\n\
 Copyright (c) 2013, Dakotah Lambert\n\
 All rights reserved.\n\
\n\
 Redistribution and use in source and binary forms, with or without\n\
 modification, are permitted provided that the following conditions\n\
 are met:\n\
\n\
 1: Redistributions of source code must retain the above copyright\n\
    notice, this list of conditions and the following disclaimer.\n\
\n\
 2: Redistributions in binary form must reproduce the above copyright\n\
    notice, this list of conditions and the following disclaimer in\n\
    the documentation and/or other materials provided with the\n\
    distribution.\n\
\n\
 3: Neither the names of copyright holders nor the names of their\n\
    contributors may be used to endors or promote products derived\n\
    from this software without specific prior written permission.\n\
\n\
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n\
  \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n\
  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS\n\
  FOR A PARTICULAR PURPOSE ARE DISCLAIMED, IN NO EVENT SHALL THE\n\
  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,\n\
  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,\n\
  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;\n\
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n\
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT\n\
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN\n\
  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE\n\
  POSSIBILITY OF SUCH DAMAGE.\n\n");
}


mdconvert/bin2smd/bin2smd.c
/*****************************************************************************
 * bin2smd: convert roms from BIN to SMD format
 * See the function ``print_license'' below for license information.
 ****************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <errno.h>

#define HEADER_SIZE 0x200
#define PERR(str) fprintf(stderr, str)

void bin2smd(FILE*, FILE*);
void print_help();
void print_license();

int
main(int argc, char* argv[])
{
	int ch = 0, hflag = 0, lflag = 0, oflag = 0;
	char output_filename[256];
	while ( (ch = getopt(argc, argv, "h?lo:")) != -1 )
	{
		switch (ch)
		{
		case 'h':
		case '?':
			hflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'o':
			oflag = 1;
			strncpy(output_filename, optarg, 255);
			output_filename[255] = '\0';
			break;
		default:
			hflag = 1;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	
	if (lflag)
	{
		print_license();
	}

	if (hflag || (argc != 1 && !lflag) || (!oflag && !lflag))
	{
		print_help();
	}
	else
	{
		if (argc == 1 && oflag)
		{
			FILE *istream, *ostream;
			istream =	fopen(argv[0], "rb");
			if (istream == NULL)
			{
				err(errno, "Could not open file %s", argv[0]);
				return EXIT_FAILURE;
			}

			ostream = fopen(output_filename, "wb");
			if (ostream == NULL)
			{
				err(errno, "Could not open file %s", output_filename);
				fclose(istream);
				return EXIT_FAILURE;
			}

			bin2smd(istream, ostream);
			fclose(istream);
			fclose(ostream);
		}
	}
	return EXIT_SUCCESS;
}


void
bin2smd(FILE *istream, FILE *ostream)
{
	ssize_t cpos = 0;
	char next_char;
	ssize_t i = 0;

	fseek(istream, 1, SEEK_SET);
	fseek(ostream, 0, SEEK_SET);
	for (i = 0; i < HEADER_SIZE; ++i)
	{
		fputc('\0', ostream);
	}

	next_char = fgetc(istream);
	while (!( feof(istream) ) && !( ferror(istream) )
	       && !( ferror(ostream) ))
	{
		fputc(next_char, ostream);
		++cpos;
		if ((cpos / 8192) % 2 == 1)
		{
			ssize_t i;
			cpos += 8192;
			for (i = 0; i < 8192; ++i) fputc('\0', ostream);
		}
		fgetc(istream);
		next_char = fgetc(istream);
	}

	cpos = 8192;
	fseek(istream, 0, SEEK_SET);
	fseek(ostream, HEADER_SIZE + 8192, SEEK_SET);
	next_char = fgetc(istream);
	while (!( feof(istream) ) && !( ferror(istream) )
	       && !( ferror(ostream) ))
	{
		fputc(next_char, ostream);
		++cpos;
		if ((cpos / 8192) % 2 == 0)
		{
			cpos += 8192;
			fseek(ostream, 8192, SEEK_CUR);
		}
		fgetc(istream);
		next_char = fgetc(istream);
	}

	fseek(ostream, 0, SEEK_SET);
	fputc((char)((cpos - 8192) / 16384), ostream);
	fputc(0x03, ostream);
	fseek(ostream, 8, SEEK_SET);
	fputs("\xAA\xBB\x06", ostream);
}


void
print_help()
{
	PERR("bin2smd Copyright (c) 2014, Dakotah Lambert.\n");
	PERR("All rights reserved.\n\n");
	PERR("usage: bin2smd [-h?l] -o outfile infile\n\n");
	PERR("  -h, -? : show this message\n");
	PERR("  -l     : display license information\n");
	PERR("  -o     : mandatory; followed by name of output file\n");
	PERR("  infile : name of file to convert\n\n");
}

void
print_license()
{
	PERR("\n bin2smd\n\
 Copyright (c) 2013, Dakotah Lambert\n\
 All rights reserved.\n\
\n\
 Redistribution and use in source and binary forms, with or without\n\
 modification, are permitted provided that the following conditions\n\
 are met:\n\
\n\
 1: Redistributions of source code must retain the above copyright\n\
    notice, this list of conditions and the following disclaimer.\n\
\n\
 2: Redistributions in binary form must reproduce the above copyright\n\
    notice, this list of conditions and the following disclaimer in\n\
    the documentation and/or other materials provided with the\n\
    distribution.\n\
\n\
 3: Neither the names of copyright holders nor the names of their\n\
    contributors may be used to endors or promote products derived\n\
    from this software without specific prior written permission.\n\
\n\
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n\
  \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n\
  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS\n\
  FOR A PARTICULAR PURPOSE ARE DISCLAIMED, IN NO EVENT SHALL THE\n\
  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,\n\
  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,\n\
  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;\n\
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n\
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT\n\
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN\n\
  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE\n\
  POSSIBILITY OF SUCH DAMAGE.\n\n");
}


mdconvert/smd2bin/smd2bin.c
/*****************************************************************************
 * smd2bin: convert roms from SMD to BIN format
 * See the function ``print_license'' below for license information.
 ****************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <errno.h>

#define HEADER_SIZE 0x200
#define PERR(str) fprintf(stderr, str)

void smd2bin(FILE*, FILE*);
void print_help();
void print_license();

int
main(int argc, char* argv[])
{
	int ch = 0, hflag = 0, lflag = 0, oflag = 0;
	char output_filename[256];
	while ( (ch = getopt(argc, argv, "h?lo:")) != -1 )
	{
		switch (ch)
		{
		case 'h':
		case '?':
			hflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'o':
			oflag = 1;
			strncpy(output_filename, optarg, 255);
			output_filename[255] = '\0';
			break;
		default:
			hflag = 1;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	
	if (lflag)
	{
		print_license();
	}

	if (hflag || (argc != 1 && !lflag) || (!oflag && !lflag))
	{
		print_help();
	}
	else
	{
		if (argc == 1 && oflag)
		{
			FILE *istream, *ostream;
			istream =	fopen(argv[0], "rb");
			if (istream == NULL)
			{
				err(errno, "Could not open file %s", argv[0]);
				return EXIT_FAILURE;
			}

			ostream = fopen(output_filename, "wb");
			if (ostream == NULL)
			{
				err(errno, "Could not open file %s", output_filename);
				fclose(istream);
				return EXIT_FAILURE;
			}

			smd2bin(istream, ostream);
			fclose(istream);
			fclose(ostream);
		}
	}
	return EXIT_SUCCESS;
}


void
smd2bin(FILE *istream, FILE *ostream)
{
	ssize_t cpos = 0;
	char next_char;
	
	fseek(istream, HEADER_SIZE, SEEK_SET);
	fseek(ostream, 0, SEEK_SET);
	next_char = fgetc(istream);
	while (!( feof(istream) ) && !( ferror(istream) )
	       && !( ferror(ostream) ))
	{
		fputc('\0', ostream);
		fputc(next_char, ostream);
		++cpos;
		if ((cpos / 8192) % 2 == 1)
		{
			cpos += 8192;
			fseek(istream, 8192, SEEK_CUR);
		}
		next_char = fgetc(istream);
	}

	cpos = 8192;
	fseek(istream, HEADER_SIZE + 8192, SEEK_SET);
	fseek(ostream, 0, SEEK_SET);
	next_char = fgetc(istream);
	while (!( feof(istream) ) && !( ferror(istream) )
	       && !( ferror(ostream) ))
	{
		fputc(next_char, ostream);
		fseek(ostream, 1, SEEK_CUR);
		++cpos;
		if ((cpos / 8192) % 2 == 0)
		{
			cpos += 8192;
			fseek(istream, 8192, SEEK_CUR);
		}
		next_char = fgetc(istream);
	}
}


void
print_help()
{
	PERR("smd2bin Copyright (c) 2013, Dakotah Lambert.\n");
	PERR("All rights reserved.\n\n");
	PERR("usage: smd2bin [-h?l] -o outfile infile\n\n");
	PERR("  -h, -? : show this message\n");
	PERR("  -l     : display license information\n");
	PERR("  -o     : mandatory; followed by name of output file\n");
	PERR("  infile : name of file to convert\n\n");
}

void
print_license()
{
	PERR("\n smd2bin\n\
 Copyright (c) 2013, Dakotah Lambert\n\
 All rights reserved.\n\
\n\
 Redistribution and use in source and binary forms, with or without\n\
 modification, are permitted provided that the following conditions\n\
 are met:\n\
\n\
 1: Redistributions of source code must retain the above copyright\n\
    notice, this list of conditions and the following disclaimer.\n\
\n\
 2: Redistributions in binary form must reproduce the above copyright\n\
    notice, this list of conditions and the following disclaimer in\n\
    the documentation and/or other materials provided with the\n\
    distribution.\n\
\n\
 3: Neither the names of copyright holders nor the names of their\n\
    contributors may be used to endors or promote products derived\n\
    from this software without specific prior written permission.\n\
\n\
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n\
  \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n\
  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS\n\
  FOR A PARTICULAR PURPOSE ARE DISCLAIMED, IN NO EVENT SHALL THE\n\
  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,\n\
  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,\n\
  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;\n\
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n\
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT\n\
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN\n\
  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE\n\
  POSSIBILITY OF SUCH DAMAGE.\n\n");
}


mbitpad/mbitpad.c
/*****************************************************************************
 * mbitpad: pad a file to an integer number of megabits
 * See the function ``print_license'' below for license information.
 ****************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <err.h>
#include <errno.h>

#define DEFAULT_CHAR ((unsigned char) 0xFF)
#define DEFAULT_SIZE 8
#define Mbit * 1024 * 1024 / 8
#define PERR(str) fprintf(stderr, str)

void pad(ssize_t, unsigned char, FILE*);
void print_help();
void print_license();

int
main(int argc, char* argv[])
{
	int ch = 0, hflag = 0, lflag = 0;
	ssize_t pad_to = DEFAULT_SIZE Mbit;
	unsigned char pad_with = DEFAULT_CHAR;
	while ( (ch = getopt(argc, argv, "h?lc:s:")) != -1 )
	{
		switch (ch)
		{
		case 'h':
		case '?':
			hflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'c':
			if (sscanf(optarg, "%hhi", &pad_with) != 1)
			{
				pad_with = DEFAULT_CHAR;
				fprintf(
					stderr,
					"Could not convert %s to a valid char, defaulting to %02hhX.",
					optarg, DEFAULT_CHAR);
			}
			break;
		case 's':
			if (sscanf(optarg, "%lud", &pad_to) != 1)
			{
				pad_to = DEFAULT_SIZE Mbit;
				fprintf(
					stderr,
					"Could not convert %s to a valid number, defaulting to %ld Mbit.",
					optarg, DEFAULT_SIZE);
			} else {
				pad_to = pad_to Mbit;
			}
			break;
		default:
			hflag = 1;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	
	if (lflag)
	{
		print_license();
	}

	if (hflag || (argc != 1 && !lflag))
	{
		print_help();
	}
	else
	{
		if (argc == 1)
		{
			FILE *ostream;
			ostream =	fopen(argv[0], "ab");
			if (ostream == NULL)
			{
				err(errno, "Could not open file %s", argv[0]);
				return EXIT_FAILURE;
			}

			pad(pad_to, pad_with, ostream);
			fclose(ostream);
		}
	}
	return EXIT_SUCCESS;
}


void
pad(ssize_t pad_to, unsigned char pad_with, FILE *ostream)
{
	ssize_t cpos = ftell(ostream);
	while (cpos < pad_to)
	{
		fputc(pad_with, ostream);
		++cpos;
	}
}


void
print_help()
{
	PERR("mbitpad Copyright (c) 2014, Dakotah Lambert.\n");
	PERR("All rights reserved.\n\n");
	PERR("usage: mbitpad [-h?l] [-c pad_char] [-s size] file\n\n");
	PERR("  -h, -? : show this message\n");
	PERR("  -l     : display license information\n");
	PERR("  -c     : ASCII value of char to pad with\n");
	PERR("  -s     : size to which the file should be padded\n");
	PERR("  file   : name of file to pad\n\n");
	PERR("Example:\n");
	PERR("mbitpad -c 0xFF -s 4 rom.bin\n");
	PERR("  pads rom.bin with character <FF> to 4 Mbit (0.5 MB)\n\n");
}

void
print_license()
{
	PERR("\n mbitpad\n\
 Copyright (c) 2014, Dakotah Lambert\n\
 All rights reserved.\n\
\n\
 Redistribution and use in source and binary forms, with or without\n\
 modification, are permitted provided that the following conditions\n\
 are met:\n\
\n\
 1: Redistributions of source code must retain the above copyright\n\
    notice, this list of conditions and the following disclaimer.\n\
\n\
 2: Redistributions in binary form must reproduce the above copyright\n\
    notice, this list of conditions and the following disclaimer in\n\
    the documentation and/or other materials provided with the\n\
    distribution.\n\
\n\
 3: Neither the names of copyright holders nor the names of their\n\
    contributors may be used to endors or promote products derived\n\
    from this software without specific prior written permission.\n\
\n\
  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n\
  \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n\
  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS\n\
  FOR A PARTICULAR PURPOSE ARE DISCLAIMED, IN NO EVENT SHALL THE\n\
  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,\n\
  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,\n\
  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;\n\
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n\
  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT\n\
  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN\n\
  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE\n\
  POSSIBILITY OF SUCH DAMAGE.\n\n");
}
